pkg/antctl/raw/upgrade/apistorage/command.go

```diff
@@ -0,0 +1,271 @@
+		crd := &apiextv1.CustomResourceDefinition{}
+		if err := k8sClient.Get(context.TODO(), client.ObjectKey{Name: crdName}, crd); err != nil {
+			return nil, fmt.Errorf("error getting CRD %q: %w", crdName, err)
+		}
+		// Skip the CRD that has only one version.
+		if len(crd.Spec.Versions) == 1 {
+			fmt.Fprintf(writer, "Skip upgrading CRD %q since it only has one version.\n", crdName)
+			continue
+		}
+		// Skip the CRD that all stored objects are in the storage version.
+		if len(crd.Status.StoredVersions) == 1 && crd.Status.StoredVersions[0] == getCRDStorageVersion(crd) {
+			fmt.Fprintf(writer, "Skip upgrading CRD %q since all stored objects are in the storage version.\n", crdName)
+			continue
+		}
+		crdsToUpgrade = append(crdsToUpgrade, crd)
+	}
+	return crdsToUpgrade, nil
+}
+
+// upgradeCRDObjects upgrades the existing objects of a CRD.
+func upgradeCRDObjects(writer io.Writer, k8sClient client.Client, crd *apiextv1.CustomResourceDefinition) error {
+	objList := &unstructured.UnstructuredList{}
+	objList.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   crd.Spec.Group,
+		Version: getCRDStorageVersion(crd),
+		Kind:    crd.Spec.Names.ListKind,
+	})
+
+	if err := k8sClient.List(context.TODO(), objList); err != nil {
+		return err
+	}
+
+	if len(objList.Items) == 0 {
+		return nil
+	}
+
+	fmt.Fprintf(writer, "Upgrading %d objects of CRD %q.\n", len(objList.Items), crd.Name)
+	for _, item := range objList.Items {
+		if err := upgradeCRDObject(writer, k8sClient, crd, item); err != nil {
+			if apierrors.IsNotFound(err) {
+				fmt.Fprintf(writer, "Skip upgrading object %s of CRD %q which is not found.\n", k8s.NamespacedName(item.GetNamespace(), item.GetName()), crd.Name)
+			} else {
+				return fmt.Errorf("error upgrading object %s of CRD %q: %w", k8s.NamespacedName(item.GetNamespace(), item.GetName()), crd.Name, err)
+			}
+		}
+	}
+	fmt.Fprintf(writer, "Successfully upgraded %d objects of CRD %q.\n", len(objList.Items), crd.Name)
+
+	return nil
+}
+
+func upgradeCRDObject(writer io.Writer, k8sClient client.Client, crd *apiextv1.CustomResourceDefinition, obj unstructured.Unstructured) error {
+	objToUpdate := &obj
+	var updateErr, getErr error
+
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		updateErr = k8sClient.Update(context.TODO(), objToUpdate)
+		// If this is a conflict error, update pointer "objToUpdate" to retry.
+		if updateErr != nil && apierrors.IsConflict(updateErr) {
+			fmt.Fprintf(writer, "Got conflict error when upgrading object %s of CRD %q, retry upgrading.\n", k8s.NamespacedName(obj.GetNamespace(), obj.GetName()), crd.Name)
+			objToUpdate = &unstructured.Unstructured{}
+			objToUpdate.SetGroupVersionKind(schema.GroupVersionKind{
+				Group:   crd.Spec.Group,
+				Version: getCRDStorageVersion(crd),
+				Kind:    crd.Spec.Names.Kind,
+			})
+			if getErr = k8sClient.Get(context.TODO(), client.ObjectKeyFromObject(&obj), objToUpdate); getErr != nil {
+				return getErr
```

Comment:
I think we can ignore the not found error here if the object is deleted during retry?

Reply:
If we got not found error, it will not retry and return the error directly.